File: commands/random.js
import { SlashCommandBuilder, EmbedBuilder, MessageFlags, ActionRowBuilder, ButtonBuilder, ButtonStyle } from 'discord.js';
import { readUserWagers, setPlayerWager, readMatchVotes, removeMatchVote, readTournamentData } from '../utils/firebase.js';
import { findNextMatch, updatePollEmbed } from '../utils/helper.js';
import { withErrorHandler, requirePlayer, requireMatches } from '../utils/command.js';
import { pickLine } from '../utils/flavor.js';

export const data = new SlashCommandBuilder()
  .setName('random')
  .setDescription('Giao số phận cho ông trời — random thay vì bị gán đội ít vote nhất');

const pendingUsers = new Set();

export const execute = withErrorHandler(async (interaction) => {
  const userId = interaction.user.id;

  if (pendingUsers.has(userId)) {
    await interaction.reply({ content: '⏳ Đang xử lý, đợi xíu...', flags: MessageFlags.Ephemeral });
    return;
  }
  pendingUsers.add(userId);

  try {
    const players = await requirePlayer(interaction, userId);
    if (!players) return;

    const allMatches = await requireMatches(interaction);
    if (!allMatches) return;

    const match = findNextMatch(allMatches);
    if (!match) {
      await interaction.reply({ content: '❌ Không có trận nào sắp tới.', flags: MessageFlags.Ephemeral });
      return;
    }

    const matchId = match.id;
    const myWagers = await readUserWagers(userId);

    if (myWagers[matchId]?.random) {
      const embed = new EmbedBuilder()
        .setTitle('⚠️  Đã kích hoạt rồi')
        .setDescription(`Random đã bật cho trận \`#${matchId}\` rồi. Nằm chờ số phận thôi!`)
        .setColor(0xFEE75C);
      await interaction.reply({ embeds: [embed], flags: MessageFlags.Ephemeral });
      return;
    }

    if (match.messageId) {
      const votes = await readMatchVotes(matchId, match.messageId);
      if (votes && votes[userId]) {
        const currentVote = votes[userId].vote;
        const row = new ActionRowBuilder().addComponents(
          new ButtonBuilder()
            .setCustomId(`random-confirm|${matchId}`)
            .setLabel('Xoá vote & random')
            .setStyle(ButtonStyle.Danger),
          new ButtonBuilder()
            .setCustomId('random-cancel')
            .setLabel('Giữ vote')
            .setStyle(ButtonStyle.Secondary),
        );

        const embed = new EmbedBuilder()
          .setTitle('⚠️  Bạn đã vote rồi')
          .setDescription(
            `Bạn đang chọn **${currentVote.toUpperCase()}** cho trận \`#${matchId}\`.\n\n` +
            'Kích hoạt random sẽ **xoá vote** hiện tại. Chắc chưa?',
          )
          .setColor(0xFEE75C);

        await interaction.reply({ embeds: [embed], components: [row], flags: MessageFlags.Ephemeral });
        return;
      }
    }

    await activateRandom(interaction, match);
  } finally {
    pendingUsers.delete(userId);
  }
});

const pendingConfirms = new Set();

export const handleRandomButton = withErrorHandler(async (interaction) => {
  if (interaction.customId === 'random-cancel') {
    const embed = new EmbedBuilder()
      .setDescription('👍 Giữ nguyên vote. Không random.')
      .setColor(0x57F287);
    await interaction.update({ embeds: [embed], components: [] });
    return;
  }

  const userId = interaction.user.id;
  if (pendingConfirms.has(userId)) {
    return;
  }
  pendingConfirms.add(userId);

  try {
    await interaction.deferUpdate();

    const [, matchIdStr] = interaction.customId.split('|');
    const matchId = parseInt(matchIdStr);

    const allMatches = await readTournamentData('matches');
    if (!allMatches) {
      await interaction.editReply({ content: '❌ Không có dữ liệu trận đấu.', embeds: [], components: [] });
      return;
    }

    const match = allMatches.find((m) => m.id === matchId);
    if (!match || Date.parse(match.date) < Date.now()) {
      await interaction.editReply({ content: '⏰ Trận đã bắt đầu hoặc không tìm thấy.', embeds: [], components: [] });
      return;
    }

    if (match.messageId) {
      await removeMatchVote(matchId, interaction.user.id, match.messageId);
      await updatePollEmbed(interaction.client, match);
    }

    await activateRandom(interaction, match, true);
  } finally {
    pendingConfirms.delete(userId);
  }
});

async function activateRandom(interaction, match, isUpdate = false) {
  await setPlayerWager(interaction.user.id, match.id, 'random');

  const randomLine = await pickLine('random');

  const embed = new EmbedBuilder()
    .setTitle('🎲  RANDOM!')
    .setDescription(
      `**${interaction.user}** ${randomLine}\n\n` +
      `⚽ **Trận #${match.id}:** ${match.home.toUpperCase()} vs ${match.away.toUpperCase()}\n\n` +
      '🎲 Nếu không vote trước giờ đá, hệ thống sẽ **random** thay vì gán đội ít vote nhất.',
    )
    .setColor(0x9B59B6)
    .setThumbnail(interaction.user.displayAvatarURL())
    .setTimestamp();

  if (isUpdate) {
    const doneEmbed = new EmbedBuilder()
      .setDescription('✅ Đã xoá vote và kích hoạt random.')
      .setColor(0x57F287);
    await interaction.editReply({ embeds: [doneEmbed], components: [] });
    await interaction.followUp({ embeds: [embed] });
  } else {
    await interaction.reply({ embeds: [embed] });
  }
}

File: commands/random.go
package commands

import (
	"context"
	"fmt"
	"sync"
	"time"

	"github.com/kdnvi/rambo-bot/bot"
	"github.com/kdnvi/rambo-bot/internal/discord"
	"github.com/kdnvi/rambo-bot/internal/firebase"
	"github.com/kdnvi/rambo-bot/internal/flavor"
)

var randomPending sync.Map
var randomConfirmPending sync.Map

func Random(ctx context.Context, b *bot.Bot, i *discord.Interaction) {
	userID := i.ActingUser().ID

	if _, loaded := randomPending.LoadOrStore(userID, struct{}{}); loaded {
		replyErr(ctx, b, i, "⏳ Đang xử lý, đợi xíu...")
		return
	}
	defer randomPending.Delete(userID)

	players, err := b.DB.ReadPlayers(ctx)
	if err != nil || players[userID] == nil {
		replyEphemeral(ctx, b, i, discord.InteractionCallbackData{
			Embeds: []discord.Embed{{
				Title:       "❌  Chưa đăng ký",
				Description: "Bạn cần `/register` trước.",
				Color:       0xED4245,
			}},
		})
		return
	}

	matches, err := b.DB.ReadMatches(ctx)
	if err != nil || len(matches) == 0 {
		replyErr(ctx, b, i, "❌ Không có dữ liệu trận đấu.")
		return
	}

	match := findNextMatch(matches)
	if match == nil {
		replyErr(ctx, b, i, "❌ Không có trận nào sắp tới.")
		return
	}

	myWagers, _ := b.DB.ReadUserWagers(ctx, userID)
	if w := myWagers[fmt.Sprintf("%d", match.ID)]; w != nil && w.Random {
		replyEphemeral(ctx, b, i, discord.InteractionCallbackData{
			Embeds: []discord.Embed{{
				Title:       "⚠️  Đã kích hoạt rồi",
				Description: fmt.Sprintf("Random đã bật cho trận `#%d` rồi. Nằm chờ số phận thôi!", match.ID),
				Color:       0xFEE75C,
			}},
		})
		return
	}

	// If user already voted, show confirmation buttons
	if match.MessageID != "" {
		votes, err := b.DB.ReadMatchVotes(ctx, match.ID, match.MessageID)
		if err == nil && votes != nil && votes[userID] != nil {
			currentVote := votes[userID].Vote
			row := discord.ActionRow(
				discord.NewButton(fmt.Sprintf("random-confirm|%d", match.ID), "Xoá vote & random", discord.ButtonDanger),
				discord.NewButton("random-cancel", "Giữ vote", discord.ButtonSecondary),
			)
			replyEphemeral(ctx, b, i, discord.InteractionCallbackData{
				Embeds: []discord.Embed{{
					Title: "⚠️  Bạn đã vote rồi",
					Description: fmt.Sprintf(
						"Bạn đang chọn **%s** cho trận `#%d`.\n\nKích hoạt random sẽ **xoá vote** hiện tại. Chắc chưa?",
						upper(currentVote), match.ID,
					),
					Color: 0xFEE75C,
				}},
				Components: []discord.Component{row},
			})
			return
		}
	}

	if err := activateRandom(ctx, b, i.ActingUser(), match.ID, match.Home, match.Away); err != nil {
		replyErr(ctx, b, i, "❌ Có lỗi xảy ra.")
		return
	}
	randomLine := flavor.PickLine(ctx, b.DB, "random")
	embed := buildRandomEmbed(i.ActingUser(), randomLine, match.ID, match.Home, match.Away)
	reply(ctx, b, i, discord.InteractionCallbackData{Embeds: []discord.Embed{embed}})
}

// HandleRandomButton handles the random-confirm and random-cancel button interactions.
func HandleRandomButton(ctx context.Context, b *bot.Bot, i *discord.Interaction) {
	if i.Data.CustomID == "random-cancel" {
		respondUpdateMessage(ctx, b, i, discord.InteractionCallbackData{
			Embeds:     []discord.Embed{{Description: "👍 Giữ nguyên vote. Không random.", Color: 0x57F287}},
			Components: []discord.Component{},
		})
		return
	}

	userID := i.ActingUser().ID
	if _, loaded := randomConfirmPending.LoadOrStore(userID, struct{}{}); loaded {
		return
	}
	defer randomConfirmPending.Delete(userID)

	deferComponentUpdate(ctx, b, i)

	var matchID int
	fmt.Sscanf(i.Data.CustomID, "random-confirm|%d", &matchID)

	matches, err := b.DB.ReadMatches(ctx)
	if err != nil {
		editOriginal(ctx, b, i, discord.EditMessagePayload{Content: "❌ Không có dữ liệu trận đấu.", Components: []discord.Component{}})
		return
	}

	var match *firebase.Match
	for idx := range matches {
		if matches[idx].ID == matchID {
			match = &matches[idx]
			break
		}
	}
	if match == nil {
		editOriginal(ctx, b, i, discord.EditMessagePayload{Content: "⏰ Trận đã bắt đầu hoặc không tìm thấy.", Components: []discord.Component{}})
		return
	}
	kickoff, _ := time.Parse(time.RFC3339, match.Date)
	if time.Now().After(kickoff) {
		editOriginal(ctx, b, i, discord.EditMessagePayload{Content: "⏰ Trận đã bắt đầu.", Components: []discord.Component{}})
		return
	}

	if match.MessageID != "" {
		_ = b.DB.RemoveMatchVote(ctx, match.ID, userID, match.MessageID)
		updatePollEmbed(ctx, b, match.ID, match.MessageID)
	}

	if err := activateRandom(ctx, b, i.ActingUser(), match.ID, match.Home, match.Away); err != nil {
		editOriginal(ctx, b, i, discord.EditMessagePayload{Content: "❌ Có lỗi xảy ra.", Components: []discord.Component{}})
		return
	}

	editOriginal(ctx, b, i, discord.EditMessagePayload{
		Embeds:     []discord.Embed{{Description: "✅ Đã xoá vote và kích hoạt random.", Color: 0x57F287}},
		Components: []discord.Component{},
	})

	randomLine := flavor.PickLine(ctx, b.DB, "random")
	embed := buildRandomEmbed(i.ActingUser(), randomLine, match.ID, match.Home, match.Away)
	followup(ctx, b, i, discord.SendMessagePayload{Embeds: []discord.Embed{embed}})
}

func activateRandom(ctx context.Context, b *bot.Bot, user *discord.User, matchID int, _, _ string) error {
	return b.DB.SetPlayerWager(ctx, user.ID, matchID, "random")
}

func buildRandomEmbed(user *discord.User, randomLine string, matchID int, home, away string) discord.Embed {
	embed := discord.Embed{
		Title: "🎲  RANDOM!",
		Description: fmt.Sprintf(
			"**<@%s>** %s\n\n⚽ **Trận #%d:** %s vs %s\n\n🎲 Nếu không vote trước giờ đá, hệ thống sẽ **random** thay vì gán đội ít vote nhất.",
			user.ID, randomLine, matchID, upper(home), upper(away),
		),
		Color:     0x9B59B6,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}
	if url := user.AvatarURL(); url != "" {
		embed.Thumbnail = &discord.EmbedImage{URL: url}
	}
	return embed
}

// respondUpdateMessage responds to a component interaction updating the original message.
func respondUpdateMessage(ctx context.Context, b *bot.Bot, i *discord.Interaction, data discord.InteractionCallbackData) {
	resp := discord.InteractionResponse{Type: discord.CallbackUpdateMessage, Data: &data}
	_ = b.REST.RespondToInteraction(ctx, i.ID, i.Token, resp)
}

// deferComponentUpdate sends a deferred update ack for a component interaction.
func deferComponentUpdate(ctx context.Context, b *bot.Bot, i *discord.Interaction) {
	resp := discord.InteractionResponse{Type: discord.CallbackDeferredMessageUpdate}
	_ = b.REST.RespondToInteraction(ctx, i.ID, i.Token, resp)
}

// editOriginal edits the original interaction message.
func editOriginal(ctx context.Context, b *bot.Bot, i *discord.Interaction, payload discord.EditMessagePayload) {
	_, _ = b.REST.EditOriginalInteractionResponse(ctx, b.AppID, i.Token, payload)
}
